package/kernel/mac80211/patches/ath/982-ath10k-qca9880-ar1a-archer-c7-wdr7500.patch

```diff
@@ -0,0 +1,916 @@
+ #define RTC_STATE_V_MASK			0x00000007
+ #define RTC_STATE_ADDRESS			0x0000
+--- a/drivers/net/wireless/ath/ath10k/mac.c
++++ b/drivers/net/wireless/ath/ath10k/mac.c
+@@ -4513,7 +4513,8 @@ void ath10k_halt(struct ath10k *ar)
+ 	ath10k_scan_finish(ar);
+ 	ath10k_peer_cleanup_all(ar);
+ 	ath10k_stop_radar_confirmation(ar);
+-	ath10k_core_stop(ar);
++	if (ar->bus_param.chip_id != 0x043200ff)
++		ath10k_core_stop(ar);
+ 	ath10k_hif_power_down(ar);
+ 
+ 	spin_lock_bh(&ar->data_lock);
+@@ -4841,17 +4842,19 @@ static int ath10k_start(struct ieee80211
+ 		goto err;
+ 	}
+ 
+-	ret = ath10k_hif_power_up(ar, ATH10K_FIRMWARE_MODE_NORMAL);
+-	if (ret) {
+-		ath10k_err(ar, "Could not init hif: %d\n", ret);
+-		goto err_off;
+-	}
+-
+-	ret = ath10k_core_start(ar, ATH10K_FIRMWARE_MODE_NORMAL,
+-				&ar->normal_mode_fw);
+-	if (ret) {
+-		ath10k_err(ar, "Could not init core: %d\n", ret);
+-		goto err_power_down;
++	if (ar->bus_param.chip_id != 0x043200ff) {
++		ret = ath10k_hif_power_up(ar, ATH10K_FIRMWARE_MODE_NORMAL);
++		if (ret) {
++			ath10k_err(ar, "Could not init hif: %d\n", ret);
++			goto err_off;
++		}
++
++		ret = ath10k_core_start(ar, ATH10K_FIRMWARE_MODE_NORMAL,
++					&ar->normal_mode_fw);
++		if (ret) {
++			ath10k_err(ar, "Could not init core: %d\n", ret);
++			goto err_power_down;
++		}
+ 	}
+ 
+ 	param = ar->wmi.pdev_param->pmf_qos;
+--- a/drivers/net/wireless/ath/ath10k/pci.c
++++ b/drivers/net/wireless/ath/ath10k/pci.c
+@@ -54,7 +54,7 @@ static const struct pci_device_id ath10k
+ 	/* PCI-E QCA988X V2 (Ubiquiti branded) */
+ 	{ PCI_VDEVICE(UBIQUITI, QCA988X_2_0_DEVICE_ID_UBNT) },
+ 
+-	{ PCI_VDEVICE(ATHEROS, QCA988X_2_0_DEVICE_ID) }, /* PCI-E QCA988X V2 */
++	{ PCI_VDEVICE(ATHEROS, QCA988X_2_0_DEVICE_ID) }, /* PCI-E QCA986X, QCA988X V1/V2 */
+ 	{ PCI_VDEVICE(ATHEROS, QCA6164_2_1_DEVICE_ID) }, /* PCI-E QCA6164 V2.1 */
+ 	{ PCI_VDEVICE(ATHEROS, QCA6174_2_1_DEVICE_ID) }, /* PCI-E QCA6174 V2.1 */
+ 	{ PCI_VDEVICE(ATHEROS, QCA99X0_2_0_DEVICE_ID) }, /* PCI-E QCA99X0 V2 */
+@@ -66,10 +66,7 @@ static const struct pci_device_id ath10k
+ };
+ 
+ static const struct ath10k_pci_supp_chip ath10k_pci_supp_chips[] = {
+-	/* QCA988X pre 2.0 chips are not supported because they need some nasty
+-	 * hacks. ath10k doesn't have them and these devices crash horribly
+-	 * because of that.
+-	 */
++	{ QCA988X_2_0_DEVICE_ID, QCA988X_HW_1_0_CHIP_ID_REV },
+ 	{ QCA988X_2_0_DEVICE_ID_UBNT, QCA988X_HW_2_0_CHIP_ID_REV },
+ 	{ QCA988X_2_0_DEVICE_ID, QCA988X_HW_2_0_CHIP_ID_REV },
+ 
+@@ -629,6 +626,7 @@ static void ath10k_pci_sleep_sync(struct
+ static void ath10k_bus_pci_write32(struct ath10k *ar, u32 offset, u32 value)
+ {
+ 	struct ath10k_pci *ar_pci = ath10k_pci_priv(ar);
++	unsigned long irq_flags;
+ 	int ret;
+ 
+ 	if (unlikely(offset + sizeof(value) > ar_pci->mem_len)) {
+@@ -644,7 +642,12 @@ static void ath10k_bus_pci_write32(struc
+ 		return;
+ 	}
+ 
++	spin_lock_irqsave(&pciwar_lock, irq_flags);
++	ioread32(ar_pci->mem + offset + 4);
```

Comment:
this one

Reply:
these reads are part of fix for EV#106293 and in case of non-bugged hardware they should not have any effect. iowrite32 still applies without shifting address.

more info:

/* BEGIN AR9888v1 WORKAROUND for EV#106293 { */

/*
 * This is a workaround for a HW issue in AR9888v1 in which there is
 * some chance that a write from Host to Target register at address X
 * may have a side effect of writing a value to X+4 (as well as the
 * requested write to X).  The value written to X+4 is whatever value
 * was read by the Host 3 reads prior.  The write to X+4 is just as
 * if software * had done it intentionally (so all effects that would
 * normally occur if that register were written do occur).
 *
 * Example1: Host tries to clear a few bits in a Copy Engine's HOST_IS
 * register (offset 0x30). As a side effect, the CE's MISC_IE register
 * (offset 0x34) is overwritten with a recently read value.
 *
 * Example2: A CE is used for Host to Target transfers, so the
 * Source Ring is maintained by the Host.  When the Host writes the
 * Source Ring Write Index, the Destination Ring Write Index is corrupted.
 *
 * The general workaround is to
 *  A) force the third read prior to a write(X) to be a read(X+4).
 *     That way, when X+4 is overwritten, it will be overwritten
 *     with the value that was there originally.
 *  B) Use a dedicated spin lock and block interrupts in order to
 *     guarantee that the above 3 reads + write occur atomically
 *     with respect to other writes from Host to Target.

Reply:
> in case of non-bugged hardware they should not have any effect.

This code is reading an unknown register and we don't know whether this register is valid on all chips supported by ath10k.
The problem is that there are a lot of new chips added after qca988xv1 support got removed. We can't guarantee that a commit didn't break old chips back then doesn't break any new things now.

Reply:
this is just a read. and for that reason it shouldn't have any effect on other pci chips. if it turns out otherwise it might reveal another hw bug, don't you think?

Reply:
Quote my reply above here:

> for example a incorrect register read may got the entire system to stuck.

For PCI devices this may be just a stuck on PCI side but for wmac devices like ipq40xx this may be a invalid memory access in kernel space, which is a kernel panic.

Reply:
is wmac interface suppose to use ath10k_bus_pci_write32 in the first place?

Reply:
> is wmac interface suppose to use ath10k_bus_pci_write32 in the first place?

I overlooked that...

But a stuck on PCI side is still broken hardware support, and the situation is the same here.

Reply:
indeed it is support, also over here